### PR TITLE
sync(sessions): vendor follow-up correctness + pricing/attribution (M6)

### PR DESCRIPTION
**Pull request by Nanako0129**
## What

**M6** of the tokscale v4.0.7 catch-up: follow-up correctness + pricing/attribution fixes. `CACHE_SCHEMA_VERSION` 21 → 22 (jcode/micode/fable/copilot changes alter cached parser output; our own counter).

| Fix | Commit | What |
|---|---|---|
| `#784` | `bc06d4ee` | **micode scan path `micode` → `mimocode`** — the old dir did not exist, so all MiMo Code usage read as empty. One-line `clients.rs` `relative` change (+ test/doc); upstream `scanner.rs` refactor **not** taken (we use the generic `*.db` glob). |
| `#747` | `c2156fea` | micode `time.created` epoch seconds-vs-ms normalization (seconds stamp no longer lands in 1970 / shrinks durations 1000×). |
| `#754` | `4bc2aa03` | jcode journal-wins over stale snapshot + tz-less ISO-8601 timestamps. `message_cache` `jcode_journal_path` relocation skipped (we have the equivalent). |
| `#724`+`#751` | `81d721fd`+`cb6ebf61` | Copilot per-message agent attribution from OTEL `gen_ai.agent.id` (sub-agents no longer collapse into one bucket) + `normalize_copilot_agent_name` + padded-id trim + cache-key parity. |
| `#762` | `afa65ed6` | `inferred_provider_from_model` maps the `fable` family to `anthropic` (bare `fable-5` stops resolving to `unknown`). Core hunk only; TUI palette half not vendored. |
| `#766` | `235230ac` | Antigravity CLI varint clamp + `saturating_add` in `TokenBreakdown::total` / `aggregate_model_usage_entries` / `positive_token_total`. Applied after #737. |
| `#662` | `63a44d7c` | **aggregator hunk only:** `calculate_summary` counts a day active on cost/messages, not only `tokens > 0`, fixing FFI `active_days` / `average_per_day`. Closes a **prior partial-vendoring gap** (the earlier sync took #662's warp `scanner.rs`/`clients.rs` parts but missed this). Warp-enable (Part B/C) still not taken (product decision). |
| `#672` | `8493048f` | Test-only: zed `XDG_DATA_HOME` env-lock guard. |

## Local wiring (self-review follow-up, commit 2)

`#724`'s `normalize_copilot_agent_name` is invoked from upstream's CLI TUI (`crates/tokscale-cli`, not vendored), so as vendored it was dead code and Copilot sub-agents surfaced with raw OTEL ids. Wired it into our `agent_bucket_key` (the #6 streaming agents report) for `client == "copilot"` — `copilot.rs`/`mod.rs` stay byte-identical to upstream; only the display bucket label changes ("github.copilot.default" → "GitHub Copilot"). Regression test added.

## Deferred / not taken

- **#752 (kiro snapshot dedup)** is entirely dependent on the `#715` globalStorage functions we do not vendor → moves to the #715 milestone.
- `#762` TUI palette half, `#784` `scanner.rs` refactor, `#662` warp-enable, `#754` `journal_path` relocation — all skipped as noted above.

## Self-review

`/code-review` (high, 3 finder angles × verify): the only actionable finding was the copilot wiring above (fixed in commit 2). Verified: all hunks faithful + coexist with our #742/#760/local patches; `#662`'s `DailyTotals.messages` is populated in our streaming path (fix is live, not a no-op); `#752` correctly not applied; saturating rewrites match upstream semantics; `#747`'s seconds/ms heuristic is a faithful upstream port (its pre-2001 edge case can't trigger for a 2024+ tool).

## Green gate

`cargo test --workspace` (27 + 949 + integ) green, `cargo clippy --workspace --all-targets` clean, `make` build, `--selftest`, `--smoke` (`top=claude-opus-4-8`, schema-22 reparse, aggregates sane, no regression).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: vendor/tokscale-core/src/sessions/antigravity_cli.rs [hunk 1]
```diff
@@ -104,11 +104,15 @@ fn parse_gen_metadata(
...
-    let output = varint_field(usage, 9).unwrap_or(0) as i64;
-    let reasoning = varint_field(usage, 10).unwrap_or(0) as i64;
+    // Clamp untrusted u64 varints into i64 (a corrupt/malicious blob could
+    // encode a value > i64::MAX, which `as i64` would wrap to a negative count)
+    // and combine with saturating_add so totals never overflow.
+    let to_i64 = |v: u64| i64::try_from(v).unwrap_or(i64::MAX);
+    let input = to_i64(varint_field(usage, 1).unwrap_or(0))
+        .saturating_add(to_i64(varint_field(usage, 2).unwrap_or(0)));
+    let cache_read = to_i64(varint_field(usage, 5).unwrap_or(0));
+    let output = to_i64(varint_field(usage, 9).unwrap_or(0));
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply vendor/tokscale-core/src/sessions/antigravity_cli.rs:114]: Fixed in b39c4c7. Both plain-`+` bucket sums in our #6 agents report — the get_agents_report sort key (lib.rs) and the FFI `total` mapper (tb_core_ffi/agents_report.rs) — now use saturating_add, so two i64::MAX-clamped buckets saturate to i64::MAX instead of overflowing (debug panic / release wrap). This also restores the documented parity with the model report total, which #766 already made saturating. Added an FFI regression test that maps an entry with i64::MAX input+output and asserts the total saturates.

File: vendor/tokscale-core/src/sessions/jcode.rs [hunk 2]
```diff
@@ -235,12 +265,33 @@ pub fn parse_jcode_file(path: &Path) -> Vec<UnifiedMessage> {
             if let Some(meta) = entry.meta {
                 context.apply_meta(meta);
             }
-            parsed.extend(parse_jcode_messages(
+            let journal_messages = parse_jcode_messages(
                 entry.append_messages,
                 &mut context,
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply vendor/tokscale-core/src/sessions/jcode.rs:270]: This journal-merge block is byte-identical to upstream #754 (junhoyeo/tokscale 4bc2aa03) — the turn-state consumption is upstream behavior, not introduced by this vendoring. The impact is a narrow undercount of the secondary turn_count metric, only when the journal's first entry corrects an *earlier* assistant id. Keeping it faithful to upstream (this is a faithful-vendor commit) and documenting it in vendor/README to report upstream rather than diverge the parser here.

File: vendor/tokscale-core/src/sessions/copilot.rs [hunk 3]
```diff
@@ -154,6 +158,19 @@ fn collect_trace_contexts(records: &[Value]) -> HashMap<String, TraceContext> {
...
+        // Trace-level agent is only a FALLBACK for records that carry no
+        // gen_ai.agent.id of their own (see candidate_from_attributes). We keep
+        // the first non-empty agent id in the trace — for Copilot CLI this is
+        // the invoke_agent span's agent, which is the right default for plain
+        // chat turns that omit the attribute. Per-record agent ids (e.g. a
+        // sub-agent turn) take precedence at attribution time, so this
+        // first-wins lock does not mis-attribute records that name their agent.
+        if context.agent_id.is_none() {
+            if let Some(agent_id) = first_non_empty_attr(attributes, &["gen_ai.agent.id"]) {
+                context.agent_id = Some(agent_id.to_string());
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply vendor/tokscale-core/src/sessions/copilot.rs:171]: This block (collect_trace_contexts) is byte-identical to upstream #724+#751 (junhoyeo/tokscale cb6ebf61). The first-non-empty-agent fallback is upstream's design — its own comment states the invoke_agent span is expected first and is the intended trace default. Pre-#724 copilot had no agent attribution at all, so this ordering edge is a limitation of the new feature, not a regression. Keeping it faithful to upstream in this faithful-vendor commit and documenting it in vendor/README to report upstream rather than diverge the parser here.

File: vendor/tokscale-core/src/sessions/antigravity_cli.rs [hunk 4]
```diff
@@ -104,11 +104,15 @@ fn parse_gen_metadata(
...
     // contradicts this, only the input total needs revisiting.
-    let input =
-        varint_field(usage, 1).unwrap_or(0) as i64 + varint_field(usage, 2).unwrap_or(0) as i64;
-    let cache_read = varint_field(usage, 5).unwrap_or(0) as i64;
-    let output = varint_field(usage, 9).unwrap_or(0) as i64;
-    let reasoning = varint_field(usage, 10).unwrap_or(0) as i64;
+    // Clamp untrusted u64 varints into i64 (a corrupt/malicious blob could
+    // encode a value > i64::MAX, which `as i64` would wrap to a negative count)
+    // and combine with saturating_add so totals never overflow.
+    let to_i64 = |v: u64| i64::try_from(v).unwrap_or(i64::MAX);
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply vendor/tokscale-core/src/sessions/antigravity_cli.rs:110]: Correct that the plain `+=` folds (aggregate_model_usage_entries, AgentAccumulator, and the daily/monthly accumulators) can still wrap in release when a corrupt DB emits two i64::MAX-clamped rows in the same bucket, before the now-saturating totals run. But the clamp value (i64::MAX) is upstream #766's and the `+=` accumulation is upstream's pre-existing fold pattern (our AgentAccumulator deliberately mirrors aggregate_model_usage_entries so the model/agent reports stay bit-identical). Making the folds saturating would diverge those core aggregators from upstream and break that documented parity, all for an adversarial/corrupt-local-DB edge that in release wraps (no crash). This is a completeness gap in #766 itself — the right fix is upstream's (saturate the folds, or clamp the varint to a summable value rather than i64::MAX). Keeping the vendored copy faithful and documenting it in vendor/README to report upstream; the user-facing total sums are already saturated (prior commit).
